src/ompl/geometric/planners/rrt/src/AOXRRTConnect.cpp

```diff
@@ -0,0 +1,536 @@
+        {
+            continue;
+        }
+
+        double cost_sample = rng_.uniformReal(0, 1);
+        auto g_hat = tgi.start ? si_->distance(rmotion->state, startState) : si_->distance(rmotion->state, goalState);
+        auto h_hat = !tgi.start ? si_->distance(rmotion->state, startState) : si_->distance(rmotion->state, goalState);
+        auto f_hat = g_hat + h_hat;
+
+        auto c_range = bestCost_.value() - f_hat;
+        if (c_range < 0)
+        {
+            continue;
+        }
+        double c_rand = (cost_sample * c_range) + g_hat;
+        rmotion->cost = c_rand;
+
+        GrowState gs = growTree(tree, tgi, rmotion);
+
+        if (gs != TRAPPED)
+        {
+            /* remember which motion was just added */
+            Motion *addedMotion = tgi.xmotion;
+
+            /* attempt to connect trees */
+
+            /* if reached, it means we used rstate directly, no need to copy again */
+            if (gs != REACHED)
+                si_->copyState(rstate, tgi.xstate);
+
+            tgi.start = !tgi.start;
+
+            cmotion->state = rmotion->state;
+            cmotion->cost = bestCost_.value() - tgi.xmotion->cost;
+            cmotion->connecting = true;
+
+            auto restored_start = false;
+            /* if initial progress cannot be done from the otherTree, restore tgi.start */
+            GrowState gsc = growTree(otherTree, tgi, cmotion);
+            if (gsc == TRAPPED)
+            {
+                tgi.start = !tgi.start;
+                restored_start = true;
+            }
+
+            while (gsc == ADVANCED)
+            {
+                gsc = growTree(otherTree, tgi, cmotion);
+            }
+
+            /* update distance between trees */
+            std::vector<Motion *> nearest_vec;
+
+            //* ------------------ ------ -------------------
+            auto *root_motion = new Motion(si_);
+
+            // could be done with restored_start == tgi.start but that isnt super clear
+            if (restored_start)
+            {
+                si_->copyState(root_motion->state, tgi.start ? goalState : startState);
+            }
+            else
+            {
+                si_->copyState(root_motion->state, !tgi.start ? goalState : startState);
+            }
+            root_motion->cost = 0;
+
+            Motion *nmotion;
+            auto rootDist = otherTree->getDistanceFunction()(addedMotion, root_motion);
+            delete root_motion;
+
+            otherTree->nearestR(addedMotion, rootDist, nearest_vec);
+            int idx = 0;
+            nmotion = nearest_vec[idx];
+            auto nearest_distance = si_->distance(addedMotion->state, nmotion->state);
+
+            while (cmotion->cost < nmotion->cost + nearest_distance && idx < (int)nearest_vec.size())
+            {
+                idx++;
+                nmotion = nearest_vec[idx];
+                nearest_distance = si_->distance(addedMotion->state, nmotion->state);
+            }
```

Comment:
Can we abstract finding neighbors in this way to a separate function? Used a few times.